draft-ietf-jsonpath-base.md

```diff
@@ -1188,17 +1188,22 @@ is between:
 is not between equal values of the same type.
 
 * a comparison using one of the operators `<`, `<=`, `>`, or `>=` yields true if and only if
-the comparison is between values of the same type which are both numbers or both strings and which satisfy the comparison:
+the comparison is between values of the same type which are both numbers, both strings, or both arrays and which satisfy the comparison:
 
     * numbers expected to interoperate as per {{Section 2.2 of -i-json (I-JSON)}} MUST compare using the normal mathematical ordering;
       numbers not expected to interoperate as per I-JSON MAY compare using an implementation specific ordering
     * the empty string compares less than any non-empty string
     * a non-empty string compares less than another non-empty string if and only if the first string starts with a
       lower Unicode scalar value than the second string or if both strings start with the same Unicode scalar value and
-      the remainder of the first string compares less than the remainder of the second string.
+      the remainder of the first string compares less than the remainder of the second string
+    * the empty array compares less than any non-empty array whose descendants do not include an object, a boolean, or `null`
+    * a non-empty array compares less than another non-empty array if and only if the first element of the first array compares
+      less than the first element of the second array or if both arrays start with the same number, string, or
+      array (whose descendants do not include an object, a boolean, or `null`) and
+      the remainder of the first array compares less than the remainder of the second array.
```

Comment:
Can this be reworded so that it's not a recursive (albeit elegant) description?

Reply:
It could be reworded to avoid the recursive definition, but I don't think the resultant text would be as clean, especially given the "non-local effect" of descendants which are objects, booleans, or `null`.

What is your concern? That the definition is hard to understand? That we might tempt implementers to use recursion? Or something else?

Reply:
My concern is that recursion is best read in code or math rather than English.  It takes a few brain cycles to understand what's going on: first that it's a recursive definition, then second to parse that definition.

I also think it could influence implementors.

Reply:
I already stated in interim meeting, that I do have objections against `</>` comparisons of different arrays.

## Arrays containing Numbers only

With arrays (seen as vectors mathematically) mostly a distinct meaning of members is implied:
* `[x,y,z]`  ... Spatial vector.
* `[r,g,b]` ... RGB color.

There is no  linear order in the plane, in the space or in any n-dimensional structure. So with the rgb color example, we won't say, that `blue` is smaller than `green` or `green` is smaller than red, i.e. `[0,0,255] < [0,255,0]`.

Of course, we could apply lexicographical ordering here, but many people would be confused by this.  In fact there is a lexicographical [ordering of vector spaces](https://en.wikipedia.org/wiki/Ordered_vector_space) defined. But from my experience in math and geometry it is of rare practical use. 

Only looking at complex numbers, which can be seen as planar vectors `[a,b]` [ordering them](https://web.maths.unsw.edu.au/~angell/articles/complexorder.pdf) is proofed as [not possible](https://math.stackexchange.com/questions/487997/total-ordering-on-complex-numbers).

## Arrays containing Strings only

The same applies to an array of strings. there must be a reason not to create a single string here.

## Arrays containing no Numbers, no Strings exclusively

I think we all agree on the fact that the relation `<` or `>` always should yield false.

Reply:
> I already stated in interim meeting, that I do have objections against `</>` comparisons of different arrays.
> ## Arrays containing Numbers only
> 
> With arrays (seen as vectors mathematically) mostly a distinct meaning of members is implied:
> 
>     * `[x,y,z]`  ... Spatial vector.
> 
>     * `[r,g,b]` ... RGB color.
> 
> 
> There is no linear order in the plane, in the space or in any n-dimensional structure. So with the rgb color example, we won't say, that `blue` is smaller than `green` or `green` is smaller than red, i.e. `[0,0,255] < [0,255,0]`.
> 
> Of course, we could apply lexicographical ordering here, but many people would be confused by this. In fact there is a lexicographical [ordering of vector spaces](https://en.wikipedia.org/wiki/Ordered_vector_space) defined. But from my experience in math and geometry it is of rare practical use.

You may be right, but why do you say many people be confused by this? Since there isn't a single obvious way of ordering arrays, they shouldn't be surprised by our choice of a lexicographic ordering.


> 
> Only looking at complex numbers, which can be seen as planar vectors `[a,b]` [ordering them](https://web.maths.unsw.edu.au/~angell/articles/complexorder.pdf) is proofed as [not possible](https://math.stackexchange.com/questions/487997/total-ordering-on-complex-numbers).

It is [possible](https://shaileshk.medium.com/an-order-on-complex-numbers-6357c8b4aec2) to define a total order on complex numbers if you regard them as merely equivalent to planar vectors. The above proof that this cannot be done uses the "field" properties of complex numbers, such as `i*i = -1`.

Reply:
> My concern is that recursion is best read in code or math rather than English. It takes a few brain cycles to understand what's going on: first that it's a recursive definition, then second to parse that definition.
> 
> I also think it could influence implementors.

Fair enough, thanks. I'm going to defer improving the text for now in case we throw out the whole PR....

Reply:
> Can this be reworded so that it's not a recursive (albeit elegant) description?

I don't think this is fundamentally possible for a format such as JSON that is essentially recursively defined.
(We can try to hide it, but boolean-expr already is recursive.)